{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1 style = \"fontsize:3rem;color:white;\">An ADAPTED Introduction to Regular Expressions in Python</h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Among the most vaulable tools available to Python programmers, are regular expressions - often shortened to as regexes. A regex is essentially a string-processing tool that allows us to pattern-match text, making them highly valuable across a variety of different applications: for everything from finding simple phrases in a mass of text, through to parsing and handling large datasets for scientific analyses.\n",
    "\n",
    "By importing the re module, pre-installed in Python, we can perform basic pattern matching using regexes. A regular expression is iself a string, that allows the user to define complex matching and replacement rules for strings, which can collectively synergise into a single-line, powerful searching, filtering and editing tool for analysing even the most expansive datasets. <p>Using minimalist syntax often referred to as regex 'special characters', we can tell the re module exactly what to match. For example a '.' special character can specifcally match a single instance of any character (text, numbers or punctuation). And a '+' can serve as a quantifier, that searches for one or more of the character that precededed it. Thus, 'a*' will search for a pattern matching anything from a single 'a' to 'aa', 'aaaaaaa' and beyond: until that pattern is interrupted by another character.  <p><b>Figure 1</b> outlines some commonly-used basic pattern matching syntax; this will be enough for us to get started matching simple alphanumeric patterns and sequences in the following exercises."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![title](Regex1.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Figure 1: </b> Table of commonly-used regular expression syntax. (Adapted from: https://www.debuggex.com/cheatsheet/regex/python)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's start by writing a simple piece of code. Complete the regular expression line to match only the numbers in the sentence, below (note the decimal point):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "sequence ='The male human genome is 6.27 Gbp. '\n",
    "\n",
    "pattern = re.compile('\\d\\.\\d\\d')\n",
    "#This line here contains our regular expression.\n",
    "#Add your special characters between the quotation marks.\n",
    "\n",
    "matches = pattern.finditer(sequence)\n",
    "#This line collates our search results into 'matches'.\n",
    "\n",
    "for match in matches:\n",
    "    print (match)\n",
    "#This for loop allows us to sequentially print each matched pattern."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, modify the code to match and output all the vowels in the following sentence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<re.Match object; span=(2, 3), match='e'>\n",
      "<re.Match object; span=(5, 6), match='a'>\n",
      "<re.Match object; span=(7, 8), match='e'>\n",
      "<re.Match object; span=(10, 11), match='u'>\n",
      "<re.Match object; span=(12, 13), match='a'>\n",
      "<re.Match object; span=(16, 17), match='e'>\n",
      "<re.Match object; span=(18, 19), match='o'>\n",
      "<re.Match object; span=(20, 21), match='e'>\n",
      "<re.Match object; span=(22, 23), match='i'>\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "sequence ='The male human genome is 6.27 Gbp. '\n",
    "\n",
    "pattern = re.compile('[a|e|i|o|u]')\n",
    "\n",
    "matches = pattern.finditer(sequence)\n",
    "\n",
    "for match in matches:\n",
    "    print (match)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that the regex returns both the match, and the 'span' - a numbered coordinate pointing to the location of the individual match in the searched string. <p>Next, let's try and work with a short nucleic acid sequence. Match all cytosine bases, and output the total number found in this sequence. Create a basic counter to tally up the number of matches and print these as a sentence ('There are xx cytosine bases in this sequence')."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 12 cytosine bases in this sequence.\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "sequence ='CTTGGGATGAAACGTTTAGCGGATAATCGTGAGATCGGACTGGCTGGATTGATCGGATTAACGGATCTGATCTGGGATCGGTAA'\n",
    "\n",
    "pattern = re.compile('C')\n",
    "\n",
    "matches = pattern.finditer(sequence)\n",
    "\n",
    "counter = 0\n",
    "\n",
    "for match in matches:\n",
    "    counter += 1\n",
    "    \n",
    "print (\"There are\", counter,\"cytosine bases in this sequence.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on this example, amend the following code to identify stop codons in the sequence given, above. Think about using an apporopriate quantifier to identify these. (TAG, TGA or TAA)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<re.Match object; span=(7, 10), match='TGA'>\n",
      "<re.Match object; span=(16, 19), match='TAG'>\n",
      "<re.Match object; span=(23, 26), match='TAA'>\n",
      "<re.Match object; span=(29, 32), match='TGA'>\n",
      "<re.Match object; span=(49, 52), match='TGA'>\n",
      "<re.Match object; span=(58, 61), match='TAA'>\n",
      "<re.Match object; span=(67, 70), match='TGA'>\n",
      "<re.Match object; span=(81, 84), match='TAA'>\n",
      "Matches = 8\n"
     ]
    }
   ],
   "source": [
    "pattern = re.compile('TGA|TAG|TAA')\n",
    "\n",
    "matches1 = pattern.finditer(sequence)\n",
    "\n",
    "counter = 0\n",
    "\n",
    "for match in matches1:\n",
    "    counter += 1\n",
    "    print (match)\n",
    "    \n",
    "print (\"Matches =\",counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many stop codons did you identify?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Answer:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above code should now identify both the number of stop codons, and all stop codons in the sequence. It does not, however, account for the different possible open reading frames. <p> For the purposes of this exercise, let's focus solely on the plus-sense ORFs. <p> The code, below, splits the sequence into these units of 3, starting from the first character of the sequence, thus taking into account the codons that comprise the +1 ORF. The output is a list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CTT', 'GGG', 'ATG', 'AAA', 'CGT', 'TTA', 'GCG', 'GAT', 'AAT', 'CGT', 'GAG', 'ATC', 'GGA', 'CTG', 'GCT', 'GGA', 'TTG', 'ATC', 'GGA', 'TTA', 'ACG', 'GAT', 'CTG', 'ATC', 'TGG', 'GAT', 'CGG', 'TAA']\n"
     ]
    }
   ],
   "source": [
    "orf1 = []\n",
    "\n",
    "n=3\n",
    "for index in range(0, len(sequence), n):\n",
    "    orf1.append(sequence[index : index + n])\n",
    "print (orf1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order for us to apply the above regex, it is easier to convert this list into a single string. We can do this by using the join() method in Python, joining each codon into a string. Using a separator of our choice, we can then make it easy to pattern match each 3-base codon (with our regex) as a separate unit, within the string. <p> Run the code in the cell below, and note the differences in the output from the cell above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTT GGG ATG AAA CGT TTA GCG GAT AAT CGT GAG ATC GGA CTG GCT GGA TTG ATC GGA TTA ACG GAT CTG ATC TGG GAT CGG TAA\n"
     ]
    }
   ],
   "source": [
    "orfjoined = ' '.join(orf1)\n",
    "print (orfjoined)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the cell below, write your own code to analyse the orfjoined string for stop codons. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<re.Match object; span=(108, 111), match='TAA'>\n",
      "\n",
      "Total +1 ORF matches: 1 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "#WRITE YOUR CODE HERE\n",
    "pattern = re.compile('TGA|TAG|TAA')\n",
    "\n",
    "matches2 = pattern.finditer(orfjoined)\n",
    "\n",
    "counter = 0\n",
    "\n",
    "for match in matches2:\n",
    "    counter += 1\n",
    "    print (match)\n",
    "    \n",
    "print ('\\nTotal +1 ORF matches:', counter, '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many do you find? What are their positions in the nucleotide sequence?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Answer:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's now take a look at the +2 ORF. Everything we have done up to this point is applicable, except we need to write code that shifts the above processes one base along the sequence. <p> An easy way to do this, is to simply create a new string with the first character of our sequence, removed. We can do this using string slicing, as follows. <p> Create a new string, whereby the first character is removed using the following syntax. Name the string you wish to slice, and enter the range you wish to return. For example, if we wish to return bases 2-8 of the sequence the code is: <p>sequence[2:8] <p> Since we only wish to remove the first character, and keep the rest of the sequence in tact, modify the code below, to do this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTG GGA TGA AAC GTT TAG CGG ATA ATC GTG AGA TCG GAC TGG CTG GAT TGA TCG GAT TAA CGG ATC TGA TCT GGG ATC GGT AA\n"
     ]
    }
   ],
   "source": [
    "orftwoseq = sequence[1:]\n",
    "orf2 = []\n",
    "\n",
    "n=3\n",
    "for index in range(0, len(orftwoseq), n):\n",
    "    orf2.append(orftwoseq[index : index + n])\n",
    "\n",
    "orfjoined = ' '.join(orf2)\n",
    "print (orfjoined)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The regex code we used earlier in the exercise, is given, below. Modify this to ascertain which of the codons in the +2 ORF are stop codons. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<re.Match object; span=(8, 11), match='TGA'>\n",
      "<re.Match object; span=(20, 23), match='TAG'>\n",
      "<re.Match object; span=(64, 67), match='TGA'>\n",
      "<re.Match object; span=(76, 79), match='TAA'>\n",
      "<re.Match object; span=(88, 91), match='TGA'>\n",
      "+2 ORF Matches = 5\n"
     ]
    }
   ],
   "source": [
    "pattern = re.compile('TGA|TAG|TAA')\n",
    "\n",
    "matches3 = pattern.finditer(orfjoined)\n",
    "\n",
    "counter = 0\n",
    "\n",
    "for match in matches3:\n",
    "    counter += 1\n",
    "    print (match)\n",
    "    \n",
    "print (\"+2 ORF Matches =\",counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many stop codons are there in the +2 open reading frame? <p> Finally, fill the cell below with the appropriate code to ascertain the number of stop codons present in the +3 ORF."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<re.Match object; span=(28, 31), match='TAA'>\n",
      "<re.Match object; span=(36, 39), match='TGA'>\n",
      "\n",
      "Total +3 ORF matches: 2\n"
     ]
    }
   ],
   "source": [
    "#WRITE CODE HERE\n",
    "\n",
    "orf3seq = sequence[2:]\n",
    "orf3 = []\n",
    "\n",
    "n=3\n",
    "for index in range(0, len(orf3seq), n):\n",
    "    orf3.append(orf3seq[index : index + n])\n",
    "\n",
    "orfjoined3 = ' '.join(orf3)\n",
    "\n",
    "pattern = re.compile('TGA|TAG|TAA')\n",
    "\n",
    "matches = pattern.finditer(orfjoined3)\n",
    "\n",
    "counter = 0\n",
    "\n",
    "for match in matches:\n",
    "    counter += 1\n",
    "    print(match)\n",
    "print ('\\nTotal +3 ORF matches:', counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many stop codons did you find in the +3 ORF? <p> Based on the stop codons counted in each of the three plus-sense ORFs, which of these reading frames is likely to translate the protein product of the gene in question?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Answer:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Special Characters and Microplastics:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following exercises aim to use Python regular expressions to sort through a microplastics dataset sampled in late 2014 across three tributaries of the river Thames. The data provides information on the site characteristics, dry weight of sediment (in grams), the number of extracted microplastic particles, and their characteristics. The data is in the form of a CSV file (comprising comma-separated values), and thus this needs to be accounted for in terms of how we read these data into Python, and pattern match them, using regular expressions. <p> The columns from left to right are: <p>Sample_ID, Site, Replicate, Size_fraction, Subsample_40_grams, Sorting_step, Other_manmade, Natural_substance, Unidentifiable.<p>Firstly, we must read the file into Python, from a file.<p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample_ID,Site,Replicate,Size_fraction,Subsample_40_grams,Sorting_step,Other_manmade,Natural_substance,Unidentifiable\n",
      "1,Lambourn,1,1-2 mm,No,Eye,0,0,1\n",
      "7,Lambourn,4,1-2 mm,Yes,Eye,1,0,0\n",
      "9,Lambourn,1,1-2 mm,No,Floated,0,0,0\n",
      "9,Lambourn,1,1-2 mm,No,Floated,0,0,1\n",
      "9,Lambourn,1,1-2 mm,No,Floated,0,0,1\n",
      "10,Lambourn,1,2-4 mm,No,Floated,0,0,1\n",
      "10,Lambourn,1,2-4 mm,No,Floated,0,0,1\n",
      "10,Lambourn,1,2-4 mm,No,Floated,0,0,1\n",
      "11,Lambourn,2,1-2 mm,No,Floated,0,0,1\n",
      "11,Lambourn,2,1-2 mm,No,Floated,0,1,0\n",
      "12,Lambourn,2,1-2 mm,Yes,Floated,1,0,0\n",
      "13,Lambourn,2,2-4 mm,No,Floated,1,0,0\n",
      "13,Lambourn,2,2-4 mm,No,Floated,0,0,1\n",
      "14,Lambourn,3,1-2 mm,No,Floated,0,0,1\n",
      "14,Lambourn,3,1-2 mm,No,Floated,0,0,1\n",
      "15,Lambourn,3,1-2 mm,Yes,Floated,0,0,1\n",
      "15,Lambourn,3,1-2 mm,Yes,Floated,0,0,1\n",
      "15,Lambourn,3,1-2 mm,Yes,Floated,0,0,1\n",
      "16,Lambourn,3,2-4 mm,No,Floated,0,0,1\n",
      "16,Lambourn,3,2-4 mm,No,Floated,0,0,1\n",
      "16,Lambourn,3,2-4 mm,No,Floated,0,0,1\n",
      "17,Lambourn,4,1-2 mm,No,Floated,0,0,1\n",
      "17,Lambourn,4,1-2 mm,No,Floated,0,0,1\n",
      "17,Lambourn,4,1-2 mm,No,Floated,1,0,0\n",
      "17,Lambourn,4,1-2 mm,No,Floated,0,0,1\n",
      "18,Lambourn,4,1-2 mm,Yes,Floated,1,0,0\n",
      "19,Lambourn,4,1-2 mm,Yes,Floated,0,0,1\n",
      "19,Lambourn,4,2-4 mm,No,Floated,1,0,0\n",
      "35,Leach,3,1-2 mm,No,Eye,0,0,1\n",
      "37,Leach,4,1-2 mm,No,Eye,0,0,1\n",
      "37,Leach,4,1-2 mm,No,Eye,0,0,0\n",
      "39,Leach,1,1-2 mm,No,Floated,0,0,1\n",
      "39,Leach,1,1-2 mm,No,Floated,0,0,1\n",
      "40,Leach,1,2-4 mm,No,Floated,0,0,1\n",
      "40,Leach,1,2-4 mm,No,Floated,0,0,1\n",
      "41,Leach,2,1-2 mm,No,Floated,0,0,1\n",
      "41,Leach,2,1-2 mm,No,Floated,1,0,0\n",
      "42,Leach,2,2-4 mm,No,Floated,0,0,1\n",
      "43,Leach,3,1-2 mm,No,Floated,0,0,1\n",
      "44,Leach,3,2-4 mm,No,Floated,0,0,1\n",
      "44,Leach,3,2-4 mm,No,Floated,1,0,0\n",
      "44,Leach,3,2-4 mm,No,Floated,0,0,1\n",
      "45,Leach,4,1-2 mm,No,Floated,0,0,1\n",
      "46,Leach,4,2-4 mm,No,Floated,1,0,0\n",
      "46,Leach,4,2-4 mm,No,Floated,0,0,1\n",
      "55,The Cut Site 1,1,1-2 mm,Yes,Eye,0,0,1\n",
      "55,The Cut Site 1,1,1-2 mm,Yes,Eye,0,0,0\n",
      "55,The Cut Site 1,1,1-2 mm,Yes,Eye,1,0,0\n",
      "56,The Cut Site 1,1,2-4 mm,No,Eye,0,0,0\n",
      "56,The Cut Site 1,1,2-4 mm,No,Eye,0,0,1\n",
      "56,The Cut Site 1,1,2-4 mm,No,Eye,1,0,0\n",
      "57,The Cut Site 1,2,1-2 mm,Yes,Eye,0,0,1\n",
      "57,The Cut Site 1,2,1-2 mm,Yes,Eye,0,0,0\n",
      "57,The Cut Site 1,2,1-2 mm,Yes,Eye,0,0,0\n",
      "57,The Cut Site 1,2,1-2 mm,Yes,Eye,1,0,0\n",
      "57,The Cut Site 1,2,1-2 mm,Yes,Eye,0,0,1\n",
      "57,The Cut Site 1,2,1-2 mm,Yes,Eye,0,0,0\n",
      "58,The Cut Site 1,2,2-4 mm,No,Eye,0,0,1\n",
      "58,The Cut Site 1,2,2-4 mm,No,Eye,0,0,1\n",
      "58,The Cut Site 1,2,2-4 mm,No,Eye,0,0,1\n",
      "59,The Cut Site 1,3,1-2 mm,Yes,Eye,1,0,0\n",
      "59,The Cut Site 1,3,1-2 mm,Yes,Eye,1,0,0\n",
      "59,The Cut Site 1,3,1-2 mm,Yes,Eye,1,0,0\n",
      "59,The Cut Site 1,3,1-2 mm,Yes,Eye,0,0,1\n",
      "59,The Cut Site 1,3,1-2 mm,Yes,Eye,0,0,0\n",
      "59,The Cut Site 1,3,1-2 mm,Yes,Eye,0,0,1\n",
      "59,The Cut Site 1,3,1-2 mm,Yes,Eye,0,0,1\n",
      "59,The Cut Site 1,3,1-2 mm,Yes,Eye,0,0,1\n",
      "60,The Cut Site 1,3,2-4 mm,No,Eye,0,0,1\n",
      "60,The Cut Site 1,3,2-4 mm,No,Eye,1,0,0\n",
      "60,The Cut Site 1,3,2-4 mm,No,Eye,1,0,0\n",
      "60,The Cut Site 1,3,2-4 mm,No,Eye,0,0,1\n",
      "60,The Cut Site 1,3,2-4 mm,No,Eye,0,0,1\n",
      "60,The Cut Site 1,3,2-4 mm,No,Eye,0,0,1\n",
      "60,The Cut Site 1,3,2-4 mm,No,Eye,0,1,0\n",
      "61,The Cut Site 1,4,1-2 mm,Yes,Eye,0,0,1\n",
      "61,The Cut Site 1,4,1-2 mm,Yes,Eye,0,0,1\n",
      "61,The Cut Site 1,4,1-2 mm,Yes,Eye,0,0,1\n",
      "61,The Cut Site 1,4,1-2 mm,Yes,Eye,0,0,1\n",
      "61,The Cut Site 1,4,1-2 mm,Yes,Eye,0,0,1\n",
      "62,The Cut Site 1,4,2-4 mm,No,Eye,0,0,0\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,0,0,1\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,0,0,1\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,0,1,0\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,1,0,0\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,0,0,0\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,0,0,1\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,1,0,0\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,0,0,0\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,0,0,1\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,0,0,1\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,0,0,1\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,0,0,1\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,0,0,1\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,1,0,0\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,0,0,1\n",
      "63,The Cut Site 1,1,1-2 mm,No,Floated,1,0,0\n",
      "64,The Cut Site 1,1,1-2 mm,Yes,Floated,0,0,1\n",
      "64,The Cut Site 1,1,1-2 mm,Yes,Floated,0,0,1\n",
      "64,The Cut Site 1,1,1-2 mm,Yes,Floated,0,0,1\n",
      "65,The Cut Site 1,1,2-4 mm,No,Floated,1,0,0\n",
      "65,The Cut Site 1,1,2-4 mm,No,Floated,1,0,0\n",
      "65,The Cut Site 1,1,2-4 mm,No,Floated,0,0,1\n",
      "65,The Cut Site 1,1,2-4 mm,No,Floated,0,0,1\n",
      "65,The Cut Site 1,1,2-4 mm,No,Floated,0,0,1\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,0,0,1\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,0,0,1\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,1,0,0\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,0,0,1\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,1,0,0\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,1,0,0\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,0,0,1\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,0,0,1\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,0,0,1\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,0,0,0\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,0,0,1\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,1,0,0\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,1,0,0\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,0,0,1\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,0,0,1\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,0,0,1\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,1,0,0\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,1,0,0\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,0,0,1\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,0,0,1\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,1,0,0\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,1,0,0\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,0,0,1\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,1,0,0\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,1,0,0\n",
      "66,The Cut Site 1,2,1-2 mm,No,Floated,1,0,0\n",
      "67,The Cut Site 1,2,1-2 mm,Yes,Floated,1,0,0\n",
      "67,The Cut Site 1,2,1-2 mm,Yes,Floated,0,0,1\n",
      "67,The Cut Site 1,2,1-2 mm,Yes,Floated,0,0,0\n",
      "67,The Cut Site 1,2,1-2 mm,Yes,Floated,0,0,1\n",
      "67,The Cut Site 1,2,1-2 mm,Yes,Floated,0,0,1\n",
      "68,The Cut Site 1,2,2-4 mm,No,Floated,0,0,0\n",
      "68,The Cut Site 1,2,2-4 mm,No,Floated,0,0,1\n",
      "68,The Cut Site 1,2,2-4 mm,No,Floated,0,0,1\n",
      "68,The Cut Site 1,2,2-4 mm,No,Floated,0,0,1\n",
      "68,The Cut Site 1,2,2-4 mm,No,Floated,0,0,1\n",
      "69,The Cut Site 1,3,1-2 mm,No,Floated,0,0,1\n",
      "69,The Cut Site 1,3,1-2 mm,No,Floated,0,0,1\n",
      "69,The Cut Site 1,3,1-2 mm,No,Floated,1,0,0\n",
      "69,The Cut Site 1,3,1-2 mm,No,Floated,0,0,0\n",
      "69,The Cut Site 1,3,1-2 mm,No,Floated,0,0,1\n",
      "69,The Cut Site 1,3,1-2 mm,No,Floated,0,0,1\n",
      "69,The Cut Site 1,3,1-2 mm,No,Floated,0,0,1\n",
      "69,The Cut Site 1,3,1-2 mm,No,Floated,1,0,0\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,0\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,1,0,0\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,0\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,0\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,1,0,0\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,0\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,1,0,0\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,0\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,0\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "70,The Cut Site 1,3,1-2 mm,Yes,Floated,0,0,1\n",
      "71,The Cut Site 1,3,2-4 mm,No,Floated,0,0,1\n",
      "71,The Cut Site 1,3,2-4 mm,No,Floated,0,0,0\n",
      "71,The Cut Site 1,3,2-4 mm,No,Floated,0,0,1\n",
      "71,The Cut Site 1,3,2-4 mm,No,Floated,0,0,1\n",
      "71,The Cut Site 1,3,2-4 mm,No,Floated,0,0,1\n",
      "71,The Cut Site 1,3,2-4 mm,No,Floated,0,0,1\n",
      "71,The Cut Site 1,3,2-4 mm,No,Floated,0,0,1\n",
      "71,The Cut Site 1,3,2-4 mm,No,Floated,0,0,1\n",
      "71,The Cut Site 1,3,2-4 mm,No,Floated,0,0,1\n",
      "72,The Cut Site 1,4,1-2 mm,No,Floated,0,0,1\n",
      "72,The Cut Site 1,4,1-2 mm,No,Floated,0,0,1\n",
      "72,The Cut Site 1,4,1-2 mm,No,Floated,0,0,0\n",
      "72,The Cut Site 1,4,1-2 mm,No,Floated,0,0,1\n",
      "72,The Cut Site 1,4,1-2 mm,No,Floated,0,0,1\n",
      "72,The Cut Site 1,4,1-2 mm,No,Floated,0,0,1\n",
      "72,The Cut Site 1,4,1-2 mm,No,Floated,0,0,1\n",
      "72,The Cut Site 1,4,1-2 mm,No,Floated,0,0,1\n",
      "72,The Cut Site 1,4,1-2 mm,No,Floated,0,0,1\n",
      "72,The Cut Site 1,4,1-2 mm,No,Floated,0,0,1\n",
      "73,The Cut Site 1,4,1-2 mm,Yes,Floated,0,0,1\n",
      "73,The Cut Site 1,4,1-2 mm,Yes,Floated,0,0,1\n",
      "73,The Cut Site 1,4,1-2 mm,Yes,Floated,0,0,0\n",
      "73,The Cut Site 1,4,1-2 mm,Yes,Floated,0,1,0\n",
      "73,The Cut Site 1,4,1-2 mm,Yes,Floated,0,0,1\n",
      "73,The Cut Site 1,4,1-2 mm,Yes,Floated,0,0,1\n",
      "74,The Cut Site 1,4,2-4 mm,No,Floated,1,0,0\n",
      "74,The Cut Site 1,4,2-4 mm,No,Floated,0,0,1\n",
      "74,The Cut Site 1,4,2-4 mm,No,Floated,0,0,1\n",
      "76,The Cut Site 1,1,1-2 mm,Yes,Unfloated,0,0,1\n",
      "76,The Cut Site 1,1,1-2 mm,Yes,Unfloated,0,0,1\n",
      "76,The Cut Site 1,1,1-2 mm,Yes,Unfloated,0,0,1\n",
      "76,The Cut Site 1,1,1-2 mm,Yes,Unfloated,0,0,1\n",
      "77,The Cut Site 1,1,2-4 mm,No,Unfloated,1,0,0\n",
      "77,The Cut Site 1,1,2-4 mm,No,Unfloated,1,0,0\n",
      "77,The Cut Site 1,1,2-4 mm,No,Unfloated,0,0,1\n",
      "77,The Cut Site 1,1,2-4 mm,No,Unfloated,0,0,1\n",
      "77,The Cut Site 1,1,2-4 mm,No,Unfloated,0,0,1\n",
      "77,The Cut Site 1,1,2-4 mm,No,Unfloated,0,0,1\n",
      "79,The Cut Site 1,2,1-2 mm,Yes,Unfloated,0,0,1\n",
      "79,The Cut Site 1,2,1-2 mm,Yes,Unfloated,1,0,0\n",
      "79,The Cut Site 1,2,1-2 mm,Yes,Unfloated,0,0,1\n",
      "79,The Cut Site 1,2,1-2 mm,Yes,Unfloated,0,0,1\n",
      "80,The Cut Site 1,2,2-4 mm,No,Unfloated,0,0,1\n",
      "80,The Cut Site 1,2,2-4 mm,No,Unfloated,0,0,1\n",
      "80,The Cut Site 1,2,2-4 mm,No,Unfloated,0,0,1\n",
      "80,The Cut Site 1,2,2-4 mm,No,Unfloated,0,0,1\n",
      "80,The Cut Site 1,2,2-4 mm,No,Unfloated,0,0,0\n",
      "80,The Cut Site 1,2,2-4 mm,No,Unfloated,1,0,0\n",
      "80,The Cut Site 1,2,2-4 mm,No,Unfloated,0,0,1\n",
      "80,The Cut Site 1,2,2-4 mm,No,Unfloated,0,0,1\n",
      "80,The Cut Site 1,2,2-4 mm,No,Unfloated,0,0,1\n",
      "82,The Cut Site 1,3,1-2 mm,Yes,Unfloated,1,0,0\n",
      "82,The Cut Site 1,3,1-2 mm,Yes,Unfloated,0,0,1\n",
      "82,The Cut Site 1,3,1-2 mm,Yes,Unfloated,0,0,1\n",
      "82,The Cut Site 1,3,1-2 mm,Yes,Unfloated,1,0,0\n",
      "82,The Cut Site 1,3,1-2 mm,Yes,Unfloated,0,0,1\n",
      "82,The Cut Site 1,3,1-2 mm,Yes,Unfloated,0,0,1\n",
      "82,The Cut Site 1,3,1-2 mm,Yes,Unfloated,0,0,0\n",
      "82,The Cut Site 1,3,1-2 mm,Yes,Unfloated,0,0,1\n",
      "82,The Cut Site 1,3,1-2 mm,Yes,Unfloated,0,0,1\n",
      "82,The Cut Site 1,3,1-2 mm,Yes,Unfloated,0,0,1\n",
      "82,The Cut Site 1,3,1-2 mm,Yes,Unfloated,0,0,1\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,0,0,1\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,0,0,1\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,0,0,1\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,0,0,1\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,0,0,1\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,0,0,1\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,1,0,0\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,0,0,1\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,0,0,1\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,0,0,1\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,1,0,0\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,0,0,1\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,1,0,0\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,0,0,1\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,0,0,0\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,1,0,0\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,0,0,1\n",
      "83,The Cut Site 1,3,2-4 mm,No,Unfloated,0,0,1\n",
      "85,The Cut Site 1,4,1-2 mm,Yes,Unfloated,0,0,0\n",
      "85,The Cut Site 1,4,1-2 mm,Yes,Unfloated,0,0,1\n",
      "85,The Cut Site 1,4,1-2 mm,Yes,Unfloated,0,0,1\n",
      "85,The Cut Site 1,4,1-2 mm,Yes,Unfloated,0,0,1\n",
      "85,The Cut Site 1,4,1-2 mm,Yes,Unfloated,0,0,1\n",
      "85,The Cut Site 1,4,1-2 mm,Yes,Unfloated,0,0,1\n",
      "85,The Cut Site 1,4,1-2 mm,Yes,Unfloated,0,0,1\n",
      "85,The Cut Site 1,4,1-2 mm,Yes,Unfloated,0,0,1\n",
      "85,The Cut Site 1,4,1-2 mm,Yes,Unfloated,1,0,0\n",
      "85,The Cut Site 1,4,1-2 mm,Yes,Unfloated,0,0,1\n",
      "85,The Cut Site 1,4,1-2 mm,Yes,Unfloated,0,0,1\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,0,0,1\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,0,0,1\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,0,0,1\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,0,0,1\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,0,0,1\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,0,0,1\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,1,0,0\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,1,0,0\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,0,0,1\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,0,0,1\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,0,0,1\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,0,0,1\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,0,0,1\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,0,1,0\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,0,0,1\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,0,0,1\n",
      "86,The Cut Site 1,4,2-4 mm,No,Unfloated,0,0,1\n",
      "87,The Cut Site 2,1,1-2 mm,No,Eye,1,0,0\n",
      "87,The Cut Site 2,1,1-2 mm,No,Eye,0,0,1\n",
      "87,The Cut Site 2,1,1-2 mm,No,Eye,0,0,1\n",
      "88,The Cut Site 2,1,2-4 mm,No,Eye,0,0,1\n",
      "89,The Cut Site 2,2,1-2 mm,No,Eye,0,0,0\n",
      "89,The Cut Site 2,2,1-2 mm,No,Eye,0,0,1\n",
      "90,The Cut Site 2,2,2-4 mm,No,Eye,0,0,0\n",
      "90,The Cut Site 2,2,2-4 mm,No,Eye,0,0,0\n",
      "91,The Cut Site 2,3,1-2 mm,Yes,Eye,0,0,1\n",
      "91,The Cut Site 2,3,1-2 mm,Yes,Eye,0,0,1\n",
      "91,The Cut Site 2,3,1-2 mm,Yes,Eye,0,0,1\n",
      "91,The Cut Site 2,3,1-2 mm,Yes,Eye,0,0,1\n",
      "91,The Cut Site 2,3,1-2 mm,Yes,Eye,0,0,1\n",
      "91,The Cut Site 2,3,1-2 mm,Yes,Eye,0,0,0\n",
      "92,The Cut Site 2,3,2-4 mm,No,Eye,0,0,0\n",
      "93,The Cut Site 2,4,1-2 mm,Yes,Eye,0,0,0\n",
      "93,The Cut Site 2,4,1-2 mm,Yes,Eye,0,0,1\n",
      "93,The Cut Site 2,4,1-2 mm,Yes,Eye,0,0,1\n",
      "94,The Cut Site 2,4,2-4 mm,No,Eye,0,1,0\n",
      "95,The Cut Site 2,1,1-2 mm,No,Floated,1,0,0\n",
      "95,The Cut Site 2,1,1-2 mm,No,Floated,0,0,0\n",
      "95,The Cut Site 2,1,1-2 mm,No,Floated,0,0,0\n",
      "96,The Cut Site 2,1,2-4 mm,No,Floated,0,0,1\n",
      "96,The Cut Site 2,1,2-4 mm,No,Floated,0,0,1\n",
      "97,The Cut Site 2,2,1-2 mm,No,Floated,0,0,1\n",
      "98,The Cut Site 2,2,2-4 mm,No,Floated,0,0,1\n",
      "98,The Cut Site 2,2,2-4 mm,No,Floated,0,0,0\n",
      "98,The Cut Site 2,2,2-4 mm,No,Floated,0,0,0\n",
      "99,The Cut Site 2,3,1-2 mm,No,Floated,0,0,0\n",
      "99,The Cut Site 2,3,1-2 mm,No,Floated,0,0,0\n",
      "99,The Cut Site 2,3,1-2 mm,No,Floated,0,0,1\n",
      "100,The Cut Site 2,3,1-2 mm,Yes,Floated,0,0,0\n",
      "100,The Cut Site 2,3,1-2 mm,Yes,Floated,0,0,1\n",
      "100,The Cut Site 2,3,1-2 mm,Yes,Floated,0,0,1\n",
      "100,The Cut Site 2,3,1-2 mm,Yes,Floated,0,0,1\n",
      "101,The Cut Site 2,3,2-4 mm,No,Floated,0,0,1\n",
      "101,The Cut Site 2,3,2-4 mm,No,Floated,0,0,1\n",
      "101,The Cut Site 2,3,2-4 mm,No,Floated,0,0,1\n",
      "101,The Cut Site 2,3,2-4 mm,No,Floated,0,0,1\n",
      "101,The Cut Site 2,3,2-4 mm,No,Floated,0,0,0\n",
      "101,The Cut Site 2,3,2-4 mm,No,Floated,0,0,1\n",
      "102,The Cut Site 2,4,1-2 mm,No,Floated,0,1,0\n",
      "102,The Cut Site 2,4,1-2 mm,No,Floated,0,0,0\n",
      "102,The Cut Site 2,4,1-2 mm,No,Floated,0,0,1\n",
      "102,The Cut Site 2,4,1-2 mm,No,Floated,1,0,0\n",
      "103,The Cut Site 2,4,1-2 mm,Yes,Floated,1,0,0\n",
      "103,The Cut Site 2,4,1-2 mm,Yes,Floated,0,0,0\n",
      "104,The Cut Site 2,4,2-4 mm,No,Floated,0,0,1\n",
      "104,The Cut Site 2,4,2-4 mm,No,Floated,0,0,1\n",
      "104,The Cut Site 2,4,2-4 mm,No,Floated,0,1,0\n",
      "110,The Cut Site 2,3,1-2 mm,Yes,Unfloated,0,0,1\n",
      "114,The Cut Site 2,4,2-4 mm,No,Unfloated,0,0,1\n"
     ]
    }
   ],
   "source": [
    "with open('ThamesMicroplastics_ParticleCharacteristics.csv', 'r') as f:\n",
    "    contents = f.read()\n",
    "print(contents)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Lambourn tributary is considerably deeper than both the Cut Site and Leach tributaries, exhibiting fast flow rates in excess of 3000 cubic feet per second (CFS), making it less favourable for studying deposited microplastics. The Cut Site and Leach tributaries are both shallower, with slower flow rates of 2-300 CFS on average. <p>Using special characters from the tables given in Figure 1, write a single regex to isolate data from all samples taken from both The Cut Site and Leach tributaries. As with the earlier exercises, keep a counter to output the number of matches returned by your regex. (Tip: You might want to use quantifier special characters to help specify your search)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of matching entries: 306\n"
     ]
    }
   ],
   "source": [
    "#WRITE YOUR CODE HERE:\n",
    "import re\n",
    "\n",
    "with open('ThamesMicroplastics_ParticleCharacteristics.csv', 'r') as f:\n",
    "    contents = f.read()\n",
    "\n",
    "pattern = re.compile('Leach|The Cut Site')\n",
    "#print (contents)\n",
    "matches = pattern.finditer(contents)\n",
    "\n",
    "counter = 0\n",
    "\n",
    "for match in matches:\n",
    "    counter += 1\n",
    "    #print (match) #Keep this to show what's matching, then comment out.\n",
    "\n",
    "print ('\\nNumber of matching entries:', counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many data entries are there for these two sites, in total?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Answer:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Copy and paste this code into the cell below, and specify the number of entries from these data that have a 2-4 mm particle size. This time, do so only for data collected at The Cut Site."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of matching entries: 106\n"
     ]
    }
   ],
   "source": [
    "#WRITE YOUR CODE HERE:\n",
    "import re\n",
    "\n",
    "with open('ThamesMicroplastics_ParticleCharacteristics.csv', 'r') as f:\n",
    "    contents = f.read()\n",
    "\n",
    "pattern = re.compile('The Cut Site.*2-4\\smm')\n",
    "\n",
    "matches = pattern.finditer(contents)\n",
    "\n",
    "counter = 0\n",
    "\n",
    "for match in matches:\n",
    "    counter += 1\n",
    "    #print (match) #Keep this to show what's matching, then comment out.\n",
    "\n",
    "print ('\\nNumber of matching entries:', counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Paste and modify your code into the cell below, this time filtering those results which were sorted by eye, and were classed as unidentifable (marked by a '1' as opposed to a '0' in the data)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of matching entries: 9\n"
     ]
    }
   ],
   "source": [
    "#WRITE YOUR CODE HERE:\n",
    "import re\n",
    "\n",
    "with open('ThamesMicroplastics_ParticleCharacteristics.csv', 'r') as f:\n",
    "    contents = f.read()\n",
    "    \n",
    "pattern = re.compile('The Cut Site.*2-4\\smm.*Eye\\W\\d\\W\\d\\W1')\n",
    "\n",
    "matches = pattern.finditer(contents)\n",
    "\n",
    "counter = 0\n",
    "\n",
    "for match in matches:\n",
    "    counter += 1\n",
    "    #print (match) #Keep this to show what's matching, then comment out.\n",
    "\n",
    "print ('\\nNumber of matching entries:', counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many data entries fit these criteria?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Answer:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Group()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Currently, your code should be matching everything between the first pattern specified in the regex, through to the last pattern. From this entire line of output, it is also possible to capture part of it, as opposed to printing the entire matched line of output. In order to achieve this, we can use '()' around the search terms in the regex, and use 'group()' to print just these captured patterns. Take the following example code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "There were 12 occurrences of C, and 21 occurrences of A.\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "sequence ='CTTGGGATGAAACGTTTAGCGGATAATCGTGAGATCGGACTGGCTGGATTGATCGGATTAACGGATCTGATCTGGGATCGGTAA'\n",
    "\n",
    "pattern = re.compile('(C)|T|(A)|G')\n",
    "\n",
    "matches = pattern.finditer(sequence)\n",
    "\n",
    "counter = 0\n",
    "counter2 = 0\n",
    "\n",
    "for match in matches:\n",
    "        if match.group(1) == 'C':\n",
    "            #print(match.group(1)) #Keep while writing, then comment out.\n",
    "            counter += 1\n",
    "        if match.group(2) == 'A':\n",
    "            #print(match.group(2)) #Keep while writing, then comment out.\n",
    "            counter2 += 1\n",
    "\n",
    "print('\\nThere were', counter, 'occurrences of C, and', counter2, 'occurrences of A.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using 'group()', we successfully isolated the bracketed portions of our regex. Note that group () also returns false values as 'None', and so there are two if loops in the code above that separate out true values, alone. (Note, feel free to try uncommenting the commented lines to see what is being matched). <p>In light of the above, incorporate 'group()' to isolate the sample ID, site name and whether substances in the samples were manmade, natural or unidentified."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample 56 - The Cut Site :  -  - unidentifiable\n",
      "Sample 58 - The Cut Site :  - natural substance - unidentifiable\n",
      "Sample 58 - The Cut Site :  -  - unidentifiable\n",
      "Sample 58 - The Cut Site : manmade -  - unidentifiable\n",
      "Sample 60 - The Cut Site :  -  - unidentifiable\n",
      "Sample 60 - The Cut Site :  -  - unidentifiable\n",
      "Sample 60 - The Cut Site :  -  - unidentifiable\n",
      "Sample 60 - The Cut Site :  -  - unidentifiable\n",
      "Sample 88 - The Cut Site :  -  - unidentifiable\n",
      "\n",
      "Total number of matching entries: 21\n"
     ]
    }
   ],
   "source": [
    "#WRITE YOUR CODE HERE:\n",
    "import re\n",
    "\n",
    "with open('ThamesMicroplastics_ParticleCharacteristics.csv', 'r') as f:\n",
    "    contents = f.read()\n",
    "    \n",
    "pattern = re.compile('(\\d*).*(The Cut Site).*2-4\\smm.*Eye,(\\d)\\W(\\d)\\W(1)')\n",
    "matches = pattern.finditer(contents)\n",
    "\n",
    "for match in matches:\n",
    "    counter += 1\n",
    "    if match.group(3) != '0':\n",
    "        category1 = 'manmade'\n",
    "    else: \n",
    "        category1 = ''\n",
    "    if match.group(4) != '0': \n",
    "        category2 = 'natural substance'\n",
    "    else: \n",
    "        category2 = ''\n",
    "    if match.group(5) != '0':\n",
    "        category3 = 'unidentifiable'\n",
    "    else: \n",
    "            category3 = ''\n",
    "    if match.group(2) == 'The Cut Site':\n",
    "        print ('Sample',match.group(1), '-', match.group(2),':', category1, '-', category2, '-', category3)\n",
    "  \n",
    "\n",
    "print ('\\nTotal number of matching entries:', counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HERV Loci"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Endogenous Retroviruses (ERVs) are among some of the most abundant sequences in mammalian genomes. Human ERVs (HERVs) are numerous, and many of them poorly described and detailed. As a result, researchers are in the process of devising a novel classification and nomenclature system to better catalogue and describe these genetic elements, but the process is long and cumbersome. <p>In order to get there, data are being collated that comprise valuable information about these HERV insertions, including their loci and genomic records. <p>In this exercise, we'll take a look at a small portion of one such dataset, and attempt to write Python regexes to capture each data criterion, so that we can parse these data down to smaller, more useful chunks. <p>As with before, the data is given as a CSV (comma-separated values) file. Run the code in the cell below, to extract the title of each column in the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record_ID,Organism,Assigned_to,Extract_start,Extract_end,Orientation,Chunk_name,Scaffold,Genome_structure\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "with open('HERV_loci.csv', 'r') as f:\n",
    "    contents = f.read()\n",
    "\n",
    "pattern = re.compile('^(.*)\\n')\n",
    "\n",
    "matches = pattern.finditer(contents)\n",
    "\n",
    "for match in matches:\n",
    "     print (match.group(1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we know what each column of data comprises, use this information to complete the code in the cell below, creating a regex that captures a value for every single criterion (column), for each row of data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45432 ERV-9 LTR-LEA\n",
      "51742\n",
      "51796 ERV-9 LTR-LEA\n",
      "52047\n",
      "52126 ERV-9 LTR-LEA\n",
      "58575\n",
      "44262 ERV-9 LTR-LEA-env-LTR\n",
      "37491\n",
      "38646 ERV-9 LTR-LEA-gag\n",
      "40592\n",
      "40688 ERV-9 LTR-LEA-gag\n",
      "42769\n",
      "44185 ERV-9 LTR-LEA-gag\n",
      "44867\n",
      "44897 ERV-9 LTR-LEA-gag\n",
      "45408\n",
      "45547 ERV-9 LTR-LEA-gag\n",
      "50355\n",
      "50368 ERV-9 LTR-LEA-gag\n",
      "51657\n",
      "52411 ERV-9 LTR-LEA-gag\n",
      "52520\n",
      "58560 ERV-9 LTR-LEA-gag\n",
      "37207\n",
      "39577 ERV-9 LTR-LEA-gag-env\n",
      "39709\n",
      "37644 ERV-9 LTR-LEA-gag-env-LTR\n",
      "40624\n",
      "40690 ERV-9 LTR-LEA-gag-env-LTR\n",
      "42858\n",
      "44181 ERV-9 LTR-LEA-gag-env-LTR\n",
      "45005\n",
      "45394 ERV-9 LTR-LEA-gag-env-LTR\n",
      "45446\n",
      "45519 ERV-9 LTR-LEA-gag-env-LTR\n",
      "49952\n",
      "50152 ERV-9 LTR-LEA-gag-env-LTR\n",
      "51697\n",
      "51797 ERV-9 LTR-LEA-gag-env-LTR\n",
      "52062\n",
      "52079 ERV-9 LTR-LEA-gag-env-LTR\n",
      "36017\n",
      "38672 ERV-9 LTR-LEA-gag-LTR\n",
      "39695\n",
      "44220 ERV-9 LTR-LEA-gag-LTR\n",
      "45402\n",
      "45583 ERV-9 LTR-LEA-gag-LTR\n",
      "47083\n",
      "49949 ERV-9 LTR-LEA-gag-LTR\n",
      "36108\n",
      "36135 ERV-9 LTR-LEA-gag-pol\n",
      "37449\n",
      "37589 ERV-9 LTR-LEA-gag-pol\n",
      "37658\n",
      "38658 ERV-9 LTR-LEA-gag-pol\n",
      "38679\n",
      "39667 ERV-9 LTR-LEA-gag-pol\n",
      "43400\n",
      "44979 ERV-9 LTR-LEA-gag-pol\n",
      "45215\n",
      "45438 ERV-9 LTR-LEA-gag-pol\n",
      "45569\n",
      "47133 ERV-9 LTR-LEA-gag-pol\n",
      "50173\n",
      "52065 ERV-9 LTR-LEA-gag-pol\n",
      "52105\n",
      "52486 ERV-9 LTR-LEA-gag-pol\n",
      "37265\n",
      "38612 ERV-9 LTR-LEA-gag-pol-env\n",
      "38635\n",
      "42763 ERV-9 LTR-LEA-gag-pol-env\n",
      "44855\n",
      "45574 ERV-9 LTR-LEA-gag-pol-env\n",
      "47116\n",
      "50352 ERV-9 LTR-LEA-gag-pol-env\n",
      "52120\n",
      "52425 ERV-9 LTR-LEA-gag-pol-env\n",
      "52442\n",
      "52497 ERV-9 LTR-LEA-gag-pol-env\n",
      "52518\n",
      "36062 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "37177\n",
      "37188 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "37222\n",
      "37229 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "37333\n",
      "37342 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "37374\n",
      "37464 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "37470\n",
      "37472 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "37488\n",
      "37519 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "37531\n",
      "37541 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "37550\n",
      "37595 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "37615\n",
      "37634 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "38614\n",
      "38616 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "39560\n",
      "39609 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "39625\n",
      "39730 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "40550\n",
      "40696 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "42760\n",
      "42764 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "42798\n",
      "42979 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "44188\n",
      "44199 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "44309\n",
      "44312 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "44449\n",
      "44846 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "44872\n",
      "44878 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "44880\n",
      "45013 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "45016\n",
      "45017 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "45338\n",
      "45347 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "45430\n",
      "45512 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "45548\n",
      "47047 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "47051\n",
      "47127 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "50212\n",
      "50229 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "50316\n",
      "50328 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "50334\n",
      "50349 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "51784\n",
      "51806 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "52023\n",
      "52050 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "52055\n",
      "52086 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "52101\n",
      "52139 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "52454\n",
      "52516 ERV-9 LTR-LEA-gag-pol-env-LTR\n",
      "36094\n",
      "36122 ERV-9 LTR-LEA-gag-pol-LTR\n",
      "37234\n",
      "37269 ERV-9 LTR-LEA-gag-pol-LTR\n",
      "37387\n",
      "37545 ERV-9 LTR-LEA-gag-pol-LTR\n",
      "38662\n",
      "39558 ERV-9 LTR-LEA-gag-pol-LTR\n",
      "39567\n",
      "39616 ERV-9 LTR-LEA-gag-pol-LTR\n",
      "39791\n",
      "42745 ERV-9 LTR-LEA-gag-pol-LTR\n",
      "42852\n",
      "42910 ERV-9 LTR-LEA-gag-pol-LTR\n",
      "42960\n",
      "42989 ERV-9 LTR-LEA-gag-pol-LTR\n",
      "43031\n",
      "44214 ERV-9 LTR-LEA-gag-pol-LTR\n",
      "44368\n",
      "45007 ERV-9 LTR-LEA-gag-pol-LTR\n",
      "45361\n",
      "45420 ERV-9 LTR-LEA-gag-pol-LTR\n",
      "45465\n",
      "50224 ERV-9 LTR-LEA-gag-pol-LTR\n",
      "50228\n",
      "50268 ERV-9 LTR-LEA-gag-pol-LTR\n",
      "51814\n",
      "52134 ERV-9 LTR-LEA-gag-pol-LTR\n",
      "42978\n",
      "45125 ERV-9 LTR-LEA-LTR\n",
      "49973\n",
      "37476 ERV-9 LTR-LEA-pol\n",
      "38605\n",
      "38618 ERV-9 LTR-LEA-pol-env-LTR\n",
      "44537\n",
      "50271 ERV-9 LTR-LEA-pol-env-LTR\n",
      "34161\n",
      "36068 ERV-9 pol\n",
      "36100\n",
      "36103 ERV-9 pol\n",
      "36118\n",
      "37211 ERV-9 pol\n",
      "37330\n",
      "37442 ERV-9 pol\n",
      "37502\n",
      "37591 ERV-9 pol\n",
      "37596\n",
      "37619 ERV-9 pol\n",
      "37637\n",
      "37645 ERV-9 pol\n",
      "38609\n",
      "38629 ERV-9 pol\n",
      "38632\n",
      "38636 ERV-9 pol\n",
      "38645\n",
      "38651 ERV-9 pol\n",
      "38657\n",
      "38666 ERV-9 pol\n",
      "39662\n",
      "39692 ERV-9 pol\n",
      "39713\n",
      "39792 ERV-9 pol\n",
      "40584\n",
      "40650 ERV-9 pol\n",
      "40669\n",
      "40672 ERV-9 pol\n",
      "40674\n",
      "40681 ERV-9 pol\n",
      "40694\n",
      "42099 ERV-9 pol\n",
      "42107\n",
      "42767 ERV-9 pol\n",
      "42768\n",
      "42804 ERV-9 pol\n",
      "42865\n",
      "42971 ERV-9 pol\n",
      "42993\n",
      "43382 ERV-9 pol\n",
      "43401\n",
      "44180 ERV-9 pol\n",
      "44202\n",
      "44229 ERV-9 pol\n",
      "44232\n",
      "44273 ERV-9 pol\n",
      "44411\n",
      "44506 ERV-9 pol\n",
      "44864\n",
      "44891 ERV-9 pol\n",
      "44981\n",
      "45350 ERV-9 pol\n",
      "45379\n",
      "45385 ERV-9 pol\n",
      "45481\n",
      "45487 ERV-9 pol\n",
      "45499\n",
      "45516 ERV-9 pol\n",
      "45551\n",
      "45565 ERV-9 pol\n",
      "45598\n",
      "47097 ERV-9 pol\n",
      "47136\n",
      "47179 ERV-9 pol\n",
      "47184\n",
      "50203 ERV-9 pol\n",
      "50252\n",
      "50280 ERV-9 pol\n",
      "50339\n",
      "50346 ERV-9 pol\n",
      "50353\n",
      "51632 ERV-9 pol\n",
      "51770\n",
      "51791 ERV-9 pol\n",
      "52017\n",
      "52053 ERV-9 pol\n",
      "52085\n",
      "52117 ERV-9 pol\n",
      "52419\n",
      "52420 ERV-9 pol\n",
      "52444\n",
      "52471 ERV-9 pol\n",
      "52494\n",
      "52511 ERV-9 pol\n",
      "52528\n",
      "34198 ERV-9 pol-env\n",
      "37444\n",
      "37581 ERV-9 pol-env\n",
      "38623\n",
      "38664 ERV-9 pol-env\n",
      "39591\n",
      "39668 ERV-9 pol-env\n",
      "42122\n",
      "44211 ERV-9 pol-env\n",
      "44531\n",
      "44932 ERV-9 pol-env\n",
      "45167\n",
      "52493 ERV-9 pol-env\n",
      "54251\n",
      "37248 ERV-9 pol-env-LTR\n",
      "37450\n",
      "39633 ERV-9 pol-env-LTR\n",
      "42863\n",
      "45143 ERV-9 pol-env-LTR\n",
      "44868\n",
      "45453 ERV-9 pol-LTR\n",
      "153043\n",
      "153050 ERV-Fc env\n",
      "153169\n",
      "153170 ERV-Fc env\n",
      "153122\n",
      "153173 ERV-Fc gag-pol-env\n",
      "153049\n",
      "153078 ERV-Fc pol\n",
      "153121\n",
      "153172 ERV-Fc pol\n",
      "68367\n",
      "68394 HERV-30 gag\n",
      "68396\n",
      "68507 HERV-30 gag\n",
      "68543\n",
      "68690 HERV-30 gag\n",
      "68528\n",
      "68650 HERV-30 gag-env\n",
      "68555\n",
      "68501 HERV-30 LEA-gag\n",
      "68540\n",
      "68544 HERV-30 LEA-gag\n",
      "68551\n",
      "68558 HERV-30 LEA-gag\n",
      "68418\n",
      "68503 HERV-30 LEA-gag-LTR\n",
      "68511\n",
      "68365 HERV-30 LTR\n",
      "68366\n",
      "68393 HERV-30 LTR\n",
      "68395\n",
      "68446 HERV-30 LTR\n",
      "68500\n",
      "68502 HERV-30 LTR\n",
      "68505\n",
      "68506 HERV-30 LTR\n",
      "68508\n",
      "68509 HERV-30 LTR\n",
      "68510\n",
      "68512 HERV-30 LTR\n",
      "68539\n",
      "68541 HERV-30 LTR\n",
      "68542\n",
      "68545 HERV-30 LTR\n",
      "68548\n",
      "68550 HERV-30 LTR\n",
      "68553\n",
      "68554 HERV-30 LTR\n",
      "68556\n",
      "68557 HERV-30 LTR\n",
      "68574\n",
      "68575 HERV-30 LTR\n",
      "68576\n",
      "68578 HERV-30 LTR\n",
      "68648\n",
      "68649 HERV-30 LTR\n",
      "68651\n",
      "68672 HERV-30 LTR\n",
      "68673\n",
      "68675 HERV-30 LTR\n",
      "68676\n",
      "68678 HERV-30 LTR\n",
      "68679\n",
      "68683 HERV-30 LTR\n",
      "68684\n",
      "68685 HERV-30 LTR\n",
      "68686\n",
      "68687 HERV-30 LTR\n",
      "68688\n",
      "68689 HERV-30 LTR\n",
      "68691\n",
      "68791 HERV-30 LTR\n",
      "68577\n",
      "68538 HERV-30 LTR-LEA-gag\n",
      "68677\n",
      "68546 HERV-30 LTR-LEA-gag-env-LTR\n",
      "68547\n",
      "68417 HERV-30 LTR-LEA-gag-LTR\n",
      "68462\n",
      "68504 HERV-30 LTR-LEA-gag-LTR\n",
      "68552\n",
      "68419 HERV-30 LTR-LEA-gag-pol-env-LTR\n",
      "68692\n",
      "153686 HERV-B7H6 env\n",
      "153689\n",
      "153691 HERV-B7H6 env\n",
      "153693\n",
      "153778 HERV-B7H6 env\n",
      "153810\n",
      "153829 HERV-B7H6 env\n",
      "153874\n",
      "154028 HERV-B7H6 env\n",
      "153606\n",
      "153608 HERV-B7H6 gag\n",
      "153649\n",
      "153688 HERV-B7H6 gag\n",
      "153713\n",
      "153776 HERV-B7H6 gag\n",
      "153777\n",
      "153779 HERV-B7H6 gag\n",
      "153781\n",
      "153786 HERV-B7H6 gag\n",
      "153812\n",
      "153813 HERV-B7H6 gag\n",
      "153814\n",
      "153828 HERV-B7H6 gag\n",
      "153830\n",
      "153840 HERV-B7H6 gag\n",
      "153844\n",
      "153873 HERV-B7H6 gag\n",
      "153966\n",
      "153969 HERV-B7H6 gag\n",
      "153970\n",
      "153972 HERV-B7H6 gag\n",
      "154013\n",
      "154014 HERV-B7H6 gag\n",
      "154016\n",
      "154026 HERV-B7H6 gag\n",
      "154029\n",
      "153842 HERV-B7H6 gag-env\n",
      "153968\n",
      "153841 HERV-B7H6 gag-pol\n",
      "153576\n",
      "153605 HERV-B7H6 LEA\n",
      "153607\n",
      "153710 HERV-B7H6 LEA\n",
      "153834\n",
      "153971 HERV-B7H6 LEA\n",
      "154007\n",
      "154025 HERV-B7H6 LEA\n",
      "153646\n",
      "153651 HERV-B7H6 LTR\n",
      "153652\n",
      "153685 HERV-B7H6 LTR\n",
      "153687\n",
      "153690 HERV-B7H6 LTR\n",
      "153692\n",
      "153709 HERV-B7H6 LTR\n",
      "153711\n",
      "153712 HERV-B7H6 LTR\n",
      "153780\n",
      "153785 HERV-B7H6 LTR\n",
      "153811\n",
      "153815 HERV-B7H6 LTR\n",
      "153818\n",
      "153819 HERV-B7H6 LTR\n",
      "153831\n",
      "153832 HERV-B7H6 LTR\n",
      "153835\n",
      "153839 HERV-B7H6 LTR\n",
      "153843\n",
      "153967 HERV-B7H6 LTR\n",
      "153973\n",
      "153974 HERV-B7H6 LTR\n",
      "154005\n",
      "154006 HERV-B7H6 LTR\n",
      "154015\n",
      "154017 HERV-B7H6 LTR\n",
      "154009\n",
      "153809 HERV-B7H6 LTR-LEA-gag-pol-env-LTR\n",
      "153647\n",
      "153648 HERV-B7H6 pol\n",
      "153650\n",
      "153816 HERV-B7H6 pol\n",
      "154008\n",
      "154010 HERV-B7H6 pol\n",
      "154024\n",
      "154027 HERV-B7H6 pol\n",
      "146663\n",
      "146666 HERV-E env\n",
      "146676\n",
      "146678 HERV-E env\n",
      "146683\n",
      "146846 HERV-E env\n",
      "146851\n",
      "146853 HERV-E env\n",
      "146863\n",
      "146865 HERV-E env\n",
      "146866\n",
      "146867 HERV-E env\n",
      "146868\n",
      "146874 HERV-E env\n",
      "146882\n",
      "146884 HERV-E env\n",
      "146981\n",
      "146986 HERV-E env\n",
      "146989\n",
      "147210 HERV-E env\n",
      "147214\n",
      "147368 HERV-E env\n",
      "147480\n",
      "147487 HERV-E env\n",
      "147529\n",
      "147646 HERV-E env\n",
      "147691\n",
      "147693 HERV-E env\n",
      "147716\n",
      "147717 HERV-E env\n",
      "147718\n",
      "147738 HERV-E env\n",
      "147740\n",
      "147864 HERV-E env\n",
      "148391\n",
      "148411 HERV-E env\n",
      "148416\n",
      "148420 HERV-E env\n",
      "148461\n",
      "148464 HERV-E env\n",
      "147696\n",
      "147091 HERV-E gag\n",
      "147105\n",
      "148424 HERV-E gag\n",
      "147369\n",
      "147704 HERV-E gag-LTR\n",
      "146664\n",
      "146982 HERV-E gag-pol\n",
      "147087\n",
      "147089 HERV-E gag-pol\n",
      "147090\n",
      "147097 HERV-E gag-pol\n",
      "147098\n",
      "147200 HERV-E gag-pol\n",
      "147201\n",
      "147209 HERV-E gag-pol\n",
      "147215\n",
      "147472 HERV-E gag-pol\n",
      "147474\n",
      "147495 HERV-E gag-pol\n",
      "147496\n",
      "147497 HERV-E gag-pol\n",
      "147631\n",
      "147632 HERV-E gag-pol\n",
      "147634\n",
      "147635 HERV-E gag-pol\n",
      "147644\n",
      "147645 HERV-E gag-pol\n",
      "147653\n",
      "147685 HERV-E gag-pol\n",
      "147699\n",
      "147700 HERV-E gag-pol\n",
      "147702\n",
      "147703 HERV-E gag-pol\n",
      "147711\n",
      "147727 HERV-E gag-pol\n",
      "147734\n",
      "147735 HERV-E gag-pol\n",
      "147736\n",
      "147743 HERV-E gag-pol\n",
      "147746\n",
      "147866 HERV-E gag-pol\n",
      "147867\n",
      "148386 HERV-E gag-pol\n",
      "148389\n",
      "148390 HERV-E gag-pol\n",
      "148422\n",
      "148423 HERV-E gag-pol\n",
      "148637\n",
      "146667 HERV-E gag-pol-env\n",
      "146859\n",
      "146860 HERV-E gag-pol-env\n",
      "146975\n",
      "146976 HERV-E gag-pol-env\n",
      "146984\n",
      "146991 HERV-E gag-pol-env\n",
      "147207\n",
      "147476 HERV-E gag-pol-env\n",
      "147492\n",
      "147689 HERV-E gag-pol-env\n",
      "147731\n",
      "148455 HERV-E gag-pol-env\n",
      "148462\n",
      "148478 HERV-E gag-pol-env\n",
      "146665\n",
      "146671 HERV-E LEA\n",
      "146674\n",
      "146675 HERV-E LEA\n",
      "146677\n",
      "146680 HERV-E LEA\n",
      "146684\n",
      "146855 HERV-E LEA\n",
      "146857\n",
      "146861 HERV-E LEA\n",
      "146869\n",
      "146988 HERV-E LEA\n",
      "147086\n",
      "147094 HERV-E LEA\n",
      "147366\n",
      "147470 HERV-E LEA\n",
      "147484\n",
      "147489 HERV-E LEA\n",
      "147528\n",
      "147530 HERV-E LEA\n",
      "147531\n",
      "147692 HERV-E LEA\n",
      "147715\n",
      "147733 HERV-E LEA\n",
      "147737\n",
      "147739 HERV-E LEA\n",
      "148241\n",
      "148243 HERV-E LEA\n",
      "148251\n",
      "148252 HERV-E LEA\n",
      "148377\n",
      "148468 HERV-E LEA\n",
      "148471\n",
      "148479 HERV-E LEA\n",
      "149011\n",
      "146661 HERV-E LEA-env\n",
      "146844\n",
      "146848 HERV-E LEA-env\n",
      "146858\n",
      "147092 HERV-E LEA-env\n",
      "147205\n",
      "147475 HERV-E LEA-env\n",
      "147478\n",
      "147483 HERV-E LEA-env\n",
      "147488\n",
      "147490 HERV-E LEA-env\n",
      "147493\n",
      "147526 HERV-E LEA-env\n",
      "147532\n",
      "147628 HERV-E LEA-env\n",
      "147636\n",
      "147663 HERV-E LEA-env\n",
      "147698\n",
      "147730 HERV-E LEA-env\n",
      "148213\n",
      "148240 HERV-E LEA-env\n",
      "148242\n",
      "148253 HERV-E LEA-env\n",
      "148382\n",
      "148383 HERV-E LEA-env\n",
      "148385\n",
      "148388 HERV-E LEA-env\n",
      "148412\n",
      "148414 HERV-E LEA-env\n",
      "148417\n",
      "148419 HERV-E LEA-env\n",
      "148472\n",
      "148475 HERV-E LEA-env\n",
      "148480\n",
      "146686 HERV-E LEA-gag-pol\n",
      "146842\n",
      "146852 HERV-E LEA-gag-pol\n",
      "146978\n",
      "147486 HERV-E LEA-gag-pol\n",
      "147655\n",
      "147705 HERV-E LEA-gag-pol\n",
      "148638\n",
      "146658 HERV-E LEA-gag-pol-env\n",
      "146682\n",
      "146854 HERV-E LEA-gag-pol-env\n",
      "146872\n",
      "147202 HERV-E LEA-gag-pol-env\n",
      "147206\n",
      "147485 HERV-E LEA-gag-pol-env\n",
      "147629\n",
      "147732 HERV-E LEA-gag-pol-env\n",
      "147742\n",
      "147749 HERV-E LEA-gag-pol-env\n",
      "147865\n",
      "148246 HERV-E LEA-gag-pol-env\n",
      "148248\n",
      "148413 HERV-E LEA-gag-pol-env\n",
      "148469\n",
      "148470 HERV-E LEA-gag-pol-env\n",
      "148476\n",
      "148477 HERV-E LEA-gag-pol-env\n",
      "146681\n",
      "147633 HERV-E LEA-gag-pol-env-LTR\n",
      "148410\n",
      "146864 HERV-E LEA-LTR\n",
      "146849\n",
      "147099 HERV-E LEA-pol\n",
      "147637\n",
      "147690 HERV-E LEA-pol\n",
      "146657\n",
      "146679 HERV-E LEA-pol-env\n",
      "146847\n",
      "146862 HERV-E LEA-pol-env\n",
      "146870\n",
      "146873 HERV-E LEA-pol-env\n",
      "146877\n",
      "146881 HERV-E LEA-pol-env\n",
      "146883\n",
      "146980 HERV-E LEA-pol-env\n",
      "147213\n",
      "147217 HERV-E LEA-pol-env\n",
      "147365\n",
      "147481 HERV-E LEA-pol-env\n",
      "147482\n",
      "147491 HERV-E LEA-pol-env\n",
      "147639\n",
      "147642 HERV-E LEA-pol-env\n",
      "147652\n",
      "147687 HERV-E LEA-pol-env\n",
      "147741\n",
      "147748 HERV-E LEA-pol-env\n",
      "148245\n",
      "148247 HERV-E LEA-pol-env\n",
      "148249\n",
      "148418 HERV-E LEA-pol-env\n",
      "148460\n",
      "148463 HERV-E LEA-pol-env\n",
      "146656\n",
      "146659 HERV-E LTR\n",
      "146660\n",
      "146668 HERV-E LTR\n",
      "146669\n",
      "146672 HERV-E LTR\n",
      "146673\n",
      "146843 HERV-E LTR\n",
      "146845\n",
      "146856 HERV-E LTR\n",
      "146871\n",
      "146875 HERV-E LTR\n",
      "146878\n",
      "147088 HERV-E LTR\n",
      "147093\n",
      "147095 HERV-E LTR\n",
      "147101\n",
      "147102 HERV-E LTR\n",
      "147107\n",
      "147109 HERV-E LTR\n",
      "147208\n",
      "147211 HERV-E LTR\n",
      "147216\n",
      "147367 HERV-E LTR\n",
      "147471\n",
      "147479 HERV-E LTR\n",
      "147525\n",
      "147534 HERV-E LTR\n",
      "147630\n",
      "147640 HERV-E LTR\n",
      "147641\n",
      "147643 HERV-E LTR\n",
      "147654\n",
      "147658 HERV-E LTR\n",
      "147661\n",
      "147664 HERV-E LTR\n",
      "147694\n",
      "147712 HERV-E LTR\n",
      "147713\n",
      "147714 HERV-E LTR\n",
      "147747\n",
      "148378 HERV-E LTR\n",
      "148379\n",
      "148380 HERV-E LTR\n",
      "148381\n",
      "148387 HERV-E LTR\n",
      "148415\n",
      "148447 HERV-E LTR\n",
      "148448\n",
      "148456 HERV-E LTR\n",
      "148458\n",
      "148459 HERV-E LTR\n",
      "148466\n",
      "148467 HERV-E LTR\n",
      "148474\n",
      "148923 HERV-E LTR\n",
      "148999\n",
      "146979 HERV-E LTR-gag-pol-env-LTR\n",
      "147103\n",
      "147662 HERV-E LTR-LEA-gag\n",
      "147686\n",
      "148211 HERV-E LTR-LEA-gag\n",
      "147688\n",
      "147100 HERV-E LTR-LEA-gag-pol-env-LTR\n",
      "147533\n",
      "148465 HERV-E LTR-LEA-gag-pol-env-LTR\n",
      "146662\n",
      "146685 HERV-E pol\n",
      "146879\n",
      "146880 HERV-E pol\n",
      "146977\n",
      "146990 HERV-E pol\n",
      "147096\n",
      "147106 HERV-E pol\n",
      "147203\n",
      "147212 HERV-E pol\n",
      "147473\n",
      "147494 HERV-E pol\n",
      "147656\n",
      "147657 HERV-E pol\n",
      "147697\n",
      "147728 HERV-E pol\n",
      "147729\n",
      "147745 HERV-E pol\n",
      "148250\n",
      "148421 HERV-E pol\n",
      "146514\n",
      "146670 HERV-E pol-env\n",
      "146850\n",
      "146876 HERV-E pol-env\n",
      "146983\n",
      "146985 HERV-E pol-env\n",
      "146987\n",
      "147104 HERV-E pol-env\n",
      "147108\n",
      "147204 HERV-E pol-env\n",
      "147477\n",
      "147527 HERV-E pol-env\n",
      "147638\n",
      "147695 HERV-E pol-env\n",
      "147726\n",
      "147744 HERV-E pol-env\n",
      "148212\n",
      "148214 HERV-E pol-env\n",
      "148244\n",
      "148384 HERV-E pol-env\n",
      "148457\n",
      "148473 HERV-E pol-env\n",
      "147701\n",
      "33688 HERV-Fb env\n",
      "33542\n",
      "33549 HERV-Fb gag\n",
      "33550\n",
      "33692 HERV-Fb gag\n",
      "33718\n",
      "33547 HERV-Fb gag-pol\n",
      "33589\n",
      "33655 HERV-Fb gag-pol\n",
      "33702\n",
      "33880 HERV-Fb gag-pol\n",
      "33901\n",
      "34029 HERV-Fb gag-pol\n",
      "33603\n",
      "33652 HERV-Fb gag-pol-env\n",
      "33720\n",
      "33514 HERV-Fb gag-pol-env-LTR\n",
      "33690\n",
      "33902 HERV-Fb gag-pol-env-LTR\n",
      "33630\n",
      "33694 HERV-Fb LEA-gag-pol\n",
      "33590\n",
      "33651 HERV-Fb LEA-gag-pol-env-LTR\n",
      "33470\n",
      "33471 HERV-Fb LTR\n",
      "33472\n",
      "33508 HERV-Fb LTR\n",
      "33513\n",
      "33546 HERV-Fb LTR\n",
      "33548\n",
      "33588 HERV-Fb LTR\n",
      "33601\n",
      "33602 HERV-Fb LTR\n",
      "33654\n",
      "33656 HERV-Fb LTR\n",
      "33657\n",
      "33665 HERV-Fb LTR\n",
      "33687\n",
      "33700 HERV-Fb LTR\n",
      "33703\n",
      "33710 HERV-Fb LTR\n",
      "33711\n",
      "33717 HERV-Fb LTR\n",
      "33719\n",
      "33721 HERV-Fb LTR\n",
      "33722\n",
      "33757 HERV-Fb LTR\n",
      "33838\n",
      "33841 HERV-Fb LTR\n",
      "33844\n",
      "33878 HERV-Fb LTR\n",
      "33881\n",
      "33883 HERV-Fb LTR\n",
      "33895\n",
      "33899 HERV-Fb LTR\n",
      "33900\n",
      "34028 HERV-Fb LTR\n",
      "33704\n",
      "33543 HERV-Fb LTR-gag-pol\n",
      "33701\n",
      "33705 HERV-Fb LTR-LEA\n",
      "33842\n",
      "33509 HERV-Fb LTR-LEA-pol\n",
      "33653\n",
      "33709 HERV-Fb LTR-pol\n",
      "33544\n",
      "33723 HERV-Fb pol\n",
      "33877\n",
      "33879 HERV-Fb pol\n",
      "33882\n",
      "33511 HERV-Fb pol-env\n",
      "33512\n",
      "33843 HERV-Fb pol-env\n",
      "33898\n",
      "33510 HERV-Fb pol-LTR\n",
      "33545\n",
      "33689 HERV-Fb pol-LTR\n",
      "33691\n",
      "33712 HERV-Fb pol-LTR\n",
      "33897\n",
      "11947 HERV-Fb-2 LEA\n",
      "11967\n",
      "12013 HERV-Fb-2 LEA\n",
      "13398\n",
      "13656 HERV-Fb-2 LEA\n",
      "13807\n",
      "14380 HERV-Fb-2 LEA\n",
      "14389\n",
      "14420 HERV-Fb-2 LEA\n",
      "14432\n",
      "14435 HERV-Fb-2 LEA\n",
      "14436\n",
      "14456 HERV-Fb-2 LEA\n",
      "14615\n",
      "14616 HERV-Fb-2 LEA\n",
      "14662\n",
      "14972 HERV-Fb-2 LEA\n",
      "15034\n",
      "15071 HERV-Fb-2 LEA\n",
      "16411\n",
      "16462 HERV-Fb-2 LEA\n",
      "19261\n",
      "19272 HERV-Fb-2 LEA\n",
      "19343\n",
      "19370 HERV-Fb-2 LEA\n",
      "20636\n",
      "20637 HERV-Fb-2 LEA\n",
      "21410\n",
      "21618 HERV-Fb-2 LEA\n",
      "21629\n",
      "27033 HERV-Fb-2 LEA\n",
      "27034\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "with open('HERV_loci.csv', 'r') as f:\n",
    "    contents = f.read()\n",
    "\n",
    "pattern = re.compile('(\\d{5,6}),(Homo_sapiens),(.[^,]*),(\\d*),(\\d*),(\\w*),(.[^,]*),(.[^,]*\\w*),(.[^,]*\\w*)')\n",
    "\n",
    "matches = pattern.finditer(contents)\n",
    "#\n",
    "for match in matches:\n",
    "     Record_ID = match.group(1)\n",
    "     Organism = match.group(2)\n",
    "     Assigned_to = match.group(3)\n",
    "     Extract_start = match.group(4)\n",
    "     Extract_end = match.group(5)\n",
    "     Orientation = match.group(6)\n",
    "     Chunk_name = match.group(7)\n",
    "     Scaffold = match.group(8)\n",
    "     Genome_structure = match.group(9)\n",
    "\n",
    "     print(Record_ID,Assigned_to,Genome_structure)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Adapting the code you just wrote, use the cell below to filter out matches that are classified as ERV-9, are in a positive orientation, and calculate the size of the described sequences. <p>Print the filtered matches as follows: <p>'Match number 48738 is a ERV-9 element, and is xxxx bases long'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Match number 45432 is a ERV-9 element, and is 1036 bases long:\n",
      " <re.Match object; span=(106, 224), match='45432,Homo_sapiens,ERV-9,146373701,146374737,posi> \n",
      "\n",
      "Match number 44185 is a ERV-9 element, and is 3269 bases long:\n",
      " <re.Match object; span=(1476, 1597), match='44185,Homo_sapiens,ERV-9,39289063,39292332,positi> \n",
      "\n",
      "Match number 44897 is a ERV-9 element, and is 2721 bases long:\n",
      " <re.Match object; span=(1707, 1827), match='44897,Homo_sapiens,ERV-9,74548527,74551248,positi> \n",
      "\n",
      "Match number 52411 is a ERV-9 element, and is 3274 bases long:\n",
      " <re.Match object; span=(2399, 2519), match='52411,Homo_sapiens,ERV-9,48664389,48667663,positi> \n",
      "\n",
      "Match number 58560 is a ERV-9 element, and is 3106 bases long:\n",
      " <re.Match object; span=(2629, 2747), match='58560,Homo_sapiens,ERV-9,408,3514,positive,hs_alt> \n",
      "\n",
      "Match number 39577 is a ERV-9 element, and is 4856 bases long:\n",
      " <re.Match object; span=(2861, 2985), match='39577,Homo_sapiens,ERV-9,80997504,81002360,positi> \n",
      "\n",
      "Match number 44181 is a ERV-9 element, and is 5901 bases long:\n",
      " <re.Match object; span=(3599, 3728), match='44181,Homo_sapiens,ERV-9,37839333,37845234,positi> \n",
      "\n",
      "Match number 45394 is a ERV-9 element, and is 5268 bases long:\n",
      " <re.Match object; span=(3846, 3974), match='45394,Homo_sapiens,ERV-9,93386023,93391291,positi> \n",
      "\n",
      "Match number 50152 is a ERV-9 element, and is 6268 bases long:\n",
      " <re.Match object; span=(4343, 4469), match='50152,Homo_sapiens,ERV-9,9598955,9605223,positive> \n",
      "\n",
      "Match number 44220 is a ERV-9 element, and is 7266 bases long:\n",
      " <re.Match object; span=(5320, 5445), match='44220,Homo_sapiens,ERV-9,92436265,92443531,positi> \n",
      "\n",
      "Match number 49949 is a ERV-9 element, and is 6858 bases long:\n",
      " <re.Match object; span=(5800, 5925), match='49949,Homo_sapiens,ERV-9,23675735,23682593,positi> \n",
      "\n",
      "Match number 45438 is a ERV-9 element, and is 4559 bases long:\n",
      " <re.Match object; span=(7233, 7359), match='45438,Homo_sapiens,ERV-9,149668820,149673379,posi> \n",
      "\n",
      "Match number 52065 is a ERV-9 element, and is 4886 bases long:\n",
      " <re.Match object; span=(7712, 7839), match='52065,Homo_sapiens,ERV-9,115272687,115277573,posi> \n",
      "\n",
      "Match number 38612 is a ERV-9 element, and is 8085 bases long:\n",
      " <re.Match object; span=(8200, 8328), match='38612,Homo_sapiens,ERV-9,10643795,10651880,positi> \n",
      "\n",
      "Match number 42763 is a ERV-9 element, and is 8845 bases long:\n",
      " <re.Match object; span=(8446, 8574), match='42763,Homo_sapiens,ERV-9,59826968,59835813,positi> \n",
      "\n",
      "Match number 52425 is a ERV-9 element, and is 7986 bases long:\n",
      " <re.Match object; span=(9186, 9314), match='52425,Homo_sapiens,ERV-9,93750933,93758919,positi> \n",
      "\n",
      "Match number 37188 is a ERV-9 element, and is 8397 bases long:\n",
      " <re.Match object; span=(9935, 10067), match='37188,Homo_sapiens,ERV-9,61331739,61340136,positi> \n",
      "\n",
      "Match number 37229 is a ERV-9 element, and is 7828 bases long:\n",
      " <re.Match object; span=(10191, 10325), match='37229,Homo_sapiens,ERV-9,126007066,126014894,posi> \n",
      "\n",
      "Match number 37464 is a ERV-9 element, and is 6631 bases long:\n",
      " <re.Match object; span=(10707, 10839), match='37464,Homo_sapiens,ERV-9,42064917,42071548,positi> \n",
      "\n",
      "Match number 37472 is a ERV-9 element, and is 6286 bases long:\n",
      " <re.Match object; span=(10961, 11093), match='37472,Homo_sapiens,ERV-9,48662318,48668604,positi> \n",
      "\n",
      "Match number 37519 is a ERV-9 element, and is 10102 bases long:\n",
      " <re.Match object; span=(11215, 11349), match='37519,Homo_sapiens,ERV-9,109507874,109517976,posi> \n",
      "\n",
      "Match number 37541 is a ERV-9 element, and is 9234 bases long:\n",
      " <re.Match object; span=(11473, 11607), match='37541,Homo_sapiens,ERV-9,134025615,134034849,posi> \n",
      "\n",
      "Match number 38616 is a ERV-9 element, and is 7526 bases long:\n",
      " <re.Match object; span=(12241, 12373), match='38616,Homo_sapiens,ERV-9,11226673,11234199,positi> \n",
      "\n",
      "Match number 39609 is a ERV-9 element, and is 7886 bases long:\n",
      " <re.Match object; span=(12495, 12629), match='39609,Homo_sapiens,ERV-9,140065628,140073514,posi> \n",
      "\n",
      "Match number 42764 is a ERV-9 element, and is 8395 bases long:\n",
      " <re.Match object; span=(13263, 13395), match='42764,Homo_sapiens,ERV-9,59959070,59967465,positi> \n",
      "\n",
      "Match number 44199 is a ERV-9 element, and is 7122 bases long:\n",
      " <re.Match object; span=(13772, 13905), match='44199,Homo_sapiens,ERV-9,63716658,63723780,positi> \n",
      "\n",
      "Match number 44846 is a ERV-9 element, and is 7507 bases long:\n",
      " <re.Match object; span=(14283, 14415), match='44846,Homo_sapiens,ERV-9,27707267,27714774,positi> \n",
      "\n",
      "Match number 44878 is a ERV-9 element, and is 7990 bases long:\n",
      " <re.Match object; span=(14537, 14669), match='44878,Homo_sapiens,ERV-9,49334531,49342521,positi> \n",
      "\n",
      "Match number 45347 is a ERV-9 element, and is 8525 bases long:\n",
      " <re.Match object; span=(15299, 15431), match='45347,Homo_sapiens,ERV-9,33428378,33436903,positi> \n",
      "\n",
      "Match number 47047 is a ERV-9 element, and is 8064 bases long:\n",
      " <re.Match object; span=(15811, 15942), match='47047,Homo_sapiens,ERV-9,4911220,4919284,positive> \n",
      "\n",
      "Match number 50229 is a ERV-9 element, and is 7391 bases long:\n",
      " <re.Match object; span=(16318, 16450), match='50229,Homo_sapiens,ERV-9,71813028,71820419,positi> \n",
      "\n",
      "Match number 52050 is a ERV-9 element, and is 6632 bases long:\n",
      " <re.Match object; span=(17337, 17470), match='52050,Homo_sapiens,ERV-9,84800221,84806853,positi> \n",
      "\n",
      "Match number 37269 is a ERV-9 element, and is 8288 bases long:\n",
      " <re.Match object; span=(18604, 18734), match='37269,Homo_sapiens,ERV-9,182071532,182079820,posi> \n",
      "\n",
      "Match number 37545 is a ERV-9 element, and is 6872 bases long:\n",
      " <re.Match object; span=(18852, 18982), match='37545,Homo_sapiens,ERV-9,141527890,141534762,posi> \n",
      "\n",
      "Match number 39558 is a ERV-9 element, and is 6191 bases long:\n",
      " <re.Match object; span=(19100, 19228), match='39558,Homo_sapiens,ERV-9,40638223,40644414,positi> \n",
      "\n",
      "Match number 39616 is a ERV-9 element, and is 6685 bases long:\n",
      " <re.Match object; span=(19346, 19476), match='39616,Homo_sapiens,ERV-9,155805705,155812390,posi> \n",
      "\n",
      "Match number 42745 is a ERV-9 element, and is 5415 bases long:\n",
      " <re.Match object; span=(19592, 19720), match='42745,Homo_sapiens,ERV-9,44855736,44861151,positi> \n",
      "\n",
      "Match number 44214 is a ERV-9 element, and is 7715 bases long:\n",
      " <re.Match object; span=(20334, 20463), match='44214,Homo_sapiens,ERV-9,81614804,81622519,positi> \n",
      "\n",
      "Match number 45420 is a ERV-9 element, and is 6788 bases long:\n",
      " <re.Match object; span=(20827, 20957), match='45420,Homo_sapiens,ERV-9,137965048,137971836,posi> \n",
      "\n",
      "Match number 50224 is a ERV-9 element, and is 5372 bases long:\n",
      " <re.Match object; span=(21077, 21205), match='50224,Homo_sapiens,ERV-9,67366938,67372310,positi> \n",
      "\n",
      "Match number 50268 is a ERV-9 element, and is 8218 bases long:\n",
      " <re.Match object; span=(21323, 21453), match='50268,Homo_sapiens,ERV-9,139062107,139070325,posi> \n",
      "\n",
      "Match number 45125 is a ERV-9 element, and is 2175 bases long:\n",
      " <re.Match object; span=(21810, 21929), match='45125,Homo_sapiens,ERV-9,4222634,4224809,positive> \n",
      "\n",
      "Match number 37476 is a ERV-9 element, and is 1383 bases long:\n",
      " <re.Match object; span=(22038, 22158), match='37476,Homo_sapiens,ERV-9,51883669,51885052,positi> \n",
      "\n",
      "Match number 38618 is a ERV-9 element, and is 8763 bases long:\n",
      " <re.Match object; span=(22270, 22398), match='38618,Homo_sapiens,ERV-9,11974985,11983748,positi> \n",
      "\n",
      "Match number 50271 is a ERV-9 element, and is 7409 bases long:\n",
      " <re.Match object; span=(22517, 22647), match='50271,Homo_sapiens,ERV-9,148733965,148741374,posi> \n",
      "\n",
      "Match number 36103 is a ERV-9 element, and is 1007 bases long:\n",
      " <re.Match object; span=(22966, 23079), match='36103,Homo_sapiens,ERV-9,24887210,24888217,positi> \n",
      "\n",
      "Match number 37211 is a ERV-9 element, and is 578 bases long:\n",
      " <re.Match object; span=(23182, 23296), match='37211,Homo_sapiens,ERV-9,101034917,101035495,posi> \n",
      "\n",
      "Match number 37442 is a ERV-9 element, and is 998 bases long:\n",
      " <re.Match object; span=(23400, 23512), match='37442,Homo_sapiens,ERV-9,20237383,20238381,positi> \n",
      "\n",
      "Match number 38629 is a ERV-9 element, and is 641 bases long:\n",
      " <re.Match object; span=(24258, 24370), match='38629,Homo_sapiens,ERV-9,15089638,15090279,positi> \n",
      "\n",
      "Match number 42099 is a ERV-9 element, and is 683 bases long:\n",
      " <re.Match object; span=(26197, 26308), match='42099,Homo_sapiens,ERV-9,3935358,3936041,positive> \n",
      "\n",
      "Match number 42767 is a ERV-9 element, and is 536 bases long:\n",
      " <re.Match object; span=(26411, 26523), match='42767,Homo_sapiens,ERV-9,65934629,65935165,positi> \n",
      "\n",
      "Match number 42804 is a ERV-9 element, and is 3570 bases long:\n",
      " <re.Match object; span=(26625, 26739), match='42804,Homo_sapiens,ERV-9,105969786,105973356,posi> \n",
      "\n",
      "Match number 43382 is a ERV-9 element, and is 878 bases long:\n",
      " <re.Match object; span=(27057, 27170), match='43382,Homo_sapiens,ERV-9,13725195,13726073,positi> \n",
      "\n",
      "Match number 44180 is a ERV-9 element, and is 2171 bases long:\n",
      " <re.Match object; span=(27273, 27386), match='44180,Homo_sapiens,ERV-9,34622357,34624528,positi> \n",
      "\n",
      "Match number 44229 is a ERV-9 element, and is 749 bases long:\n",
      " <re.Match object; span=(27489, 27604), match='44229,Homo_sapiens,ERV-9,103997440,103998189,posi> \n",
      "\n",
      "Match number 44506 is a ERV-9 element, and is 959 bases long:\n",
      " <re.Match object; span=(27924, 28037), match='44506,Homo_sapiens,ERV-9,48549935,48550894,positi> \n",
      "\n",
      "Match number 44891 is a ERV-9 element, and is 2309 bases long:\n",
      " <re.Match object; span=(28139, 28251), match='44891,Homo_sapiens,ERV-9,68689014,68691323,positi> \n",
      "\n",
      "Match number 45350 is a ERV-9 element, and is 707 bases long:\n",
      " <re.Match object; span=(28353, 28465), match='45350,Homo_sapiens,ERV-9,36115430,36116137,positi> \n",
      "\n",
      "Match number 45385 is a ERV-9 element, and is 677 bases long:\n",
      " <re.Match object; span=(28567, 28679), match='45385,Homo_sapiens,ERV-9,82714731,82715408,positi> \n",
      "\n",
      "Match number 47097 is a ERV-9 element, and is 665 bases long:\n",
      " <re.Match object; span=(29431, 29544), match='47097,Homo_sapiens,ERV-9,62177320,62177985,positi> \n",
      "\n",
      "Match number 50203 is a ERV-9 element, and is 488 bases long:\n",
      " <re.Match object; span=(29863, 29975), match='50203,Homo_sapiens,ERV-9,48715939,48716427,positi> \n",
      "\n",
      "Match number 50280 is a ERV-9 element, and is 1898 bases long:\n",
      " <re.Match object; span=(30079, 30193), match='50280,Homo_sapiens,ERV-9,166948903,166950801,posi> \n",
      "\n",
      "Match number 51632 is a ERV-9 element, and is 1571 bases long:\n",
      " <re.Match object; span=(30509, 30621), match='51632,Homo_sapiens,ERV-9,12208512,12210083,positi> \n",
      "\n",
      "Match number 52053 is a ERV-9 element, and is 1862 bases long:\n",
      " <re.Match object; span=(30938, 31053), match='52053,Homo_sapiens,ERV-9,101664613,101666475,posi> \n",
      "\n",
      "Match number 52420 is a ERV-9 element, and is 2732 bases long:\n",
      " <re.Match object; span=(31371, 31483), match='52420,Homo_sapiens,ERV-9,75754975,75757707,positi> \n",
      "\n",
      "Match number 44211 is a ERV-9 element, and is 2897 bases long:\n",
      " <re.Match object; span=(32913, 33030), match='44211,Homo_sapiens,ERV-9,78634019,78636916,positi> \n",
      "\n",
      "Match number 44932 is a ERV-9 element, and is 1913 bases long:\n",
      " <re.Match object; span=(33137, 33255), match='44932,Homo_sapiens,ERV-9,124710081,124711994,posi> \n",
      "\n",
      "Match number 37248 is a ERV-9 element, and is 4323 bases long:\n",
      " <re.Match object; span=(33583, 33705), match='37248,Homo_sapiens,ERV-9,153030400,153034723,posi> \n",
      "\n",
      "Match number 39633 is a ERV-9 element, and is 2647 bases long:\n",
      " <re.Match object; span=(33815, 33937), match='39633,Homo_sapiens,ERV-9,186179546,186182193,posi> \n",
      "\n",
      "Match number 45143 is a ERV-9 element, and is 4072 bases long:\n",
      " <re.Match object; span=(34049, 34170), match='45143,Homo_sapiens,ERV-9,28786309,28790381,positi> \n",
      "\n",
      "Match number 45453 is a ERV-9 element, and is 4787 bases long:\n",
      " <re.Match object; span=(34276, 34395), match='45453,Homo_sapiens,ERV-9,161901686,161906473,posi> \n",
      "\n"
     ]
    }
   ],
   "source": [
    "#WRITE YOUR CODE HERE\n",
    "import re\n",
    "\n",
    "with open('HERV_loci.csv', 'r') as f:\n",
    "    contents = f.read()\n",
    "    \n",
    "pattern = re.compile('(\\d{5,6}),(.[^,]*),(.[^,]*),(\\d*),(\\d*),(\\w*),(.[^,]*),(.[^,]*\\w*),(.[^,]*\\w*)')\n",
    "\n",
    "matches = pattern.finditer(contents)\n",
    "counter = 0\n",
    "for match in matches:\n",
    "     # print (match)\n",
    "     Record_ID = match.group(1)\n",
    "     Organism = match.group(2)\n",
    "     Assigned_to = match.group(3)\n",
    "     Extract_start = match.group(4)\n",
    "     Extract_end = match.group(5)\n",
    "     Orientation = match.group(6)\n",
    "     Chunk_name = match.group(7)\n",
    "     Scaffold = match.group(8)\n",
    "     Genome_structure = match.group(9)\n",
    "\n",
    "     if Assigned_to == 'ERV-9':\n",
    "          if Orientation == 'positive':\n",
    "               Extract_length = int(Extract_end) - int(Extract_start)\n",
    "               print ('Match number',Record_ID,'is a', Assigned_to, 'element, and is', Extract_length, 'bases long:\\n', match,'\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Having scratched the surface of using regular expressions in Python to handle data, it is obvious to see their power and usefulness as applied to biological sciences. Should you wish to delve into these further, below are some useful online resources: <p>https://docs.python.org/3/howto/regex.html <p>https://realpython.com/regex-python/"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
